service/history/workflow/mutable_state_impl.go

```diff
@@ -98,6 +98,13 @@ const (
 	int64SizeBytes = 8
 )
 
+var (
+	// Scheduled tasks with timestamp after this will not be created.
+	// Those tasks are too far in the future and pratically never fire and just consume storage space.
+	// NOTE: this value is less than time.Now() + timer.MaxAllowedTimer so that no capped timers will be created.
+	maxScheduledTaskTimestamp = time.Date(2100, 1, 1, 0, 0, 0, 0, time.UTC)
```

Comment:
instead of a fixed point in time, can we check the duration from the fire time and now, and skip creating it if it is exceeding a limit (say 100 years).

Reply:
Task are defined in terms of timestamp so that will need an additional step for converting it back to duration and then compare. I also want to make sure all capped timers are dropped as well, which means we need another limit lower than 100 years...
Using a timestamp feels much simpler to me. But let me know if you have strong opinion on it.

Reply:
Isn't it just task.GetVisibilityTime().Sub(time.Now()) > maxTimerDuration

Reply:
yeah it is and it's my first version. I just don't see the why it's necessary to call time.Now() and Sub() every AddTasks() if all it takes is a comparison. The duration limit is magic number anyway.

I can change to the approach you suggested but I will defined other duration limit lower than the one we capped on frontend, say 99 years.

Reply:
done